File: day16/aoc16.py
from queue import PriorityQueue

L = ("L", 1000)
R = ("R", 1000)
F = ("F", 1)

actions = [L, R, F]

N = (-1, 0)
E = (0, 1)
S = (1, 0)
W = (0, -1)

turns = {
    (N, L): W,
    (N, R): E,
    (E, L): N,
    (E, R): S,
    (S, L): E,
    (S, R): W,
    (W, L): S,
    (W, R): N,
}


def get_map(path):
    with open(path, "r") as file:
        lines = file.readlines()
        return lines


def get_start_and_goal(grid):
    goal_found = False
    start_found = False
    goal_x = goal_y = start_x = start_y = 0
    for x, row in enumerate(grid):
        for y, elem in enumerate(row):
            if elem == "E":
                goal_x, goal_y = x, y
                goal_found = True
            if elem == "S":
                start_x, start_y = x, y
                start_found = True
            if goal_found and start_found:
                return (start_x, start_y), (goal_x, goal_y)


class State:
    def __init__(self, x, y, dir):
        self.x = x
        self.y = y
        self.dir = dir

    def __str__(self):
        return f"Pos: ({self.x}, {self.y}), Dir: {self.dir}"

    def __repr__(self):
        return self.__str__()


def clear(state, dir):
    return grid[state.x + dir[0]][state.y + dir[1]] == "."


def valid(state, action):
    if action == F:
        return clear(state, state.dir)
    return clear(state, turns[(state.dir, action)])


def update_state(state, action):
    if action == F:
        state.x += state.dir[0]
        state.y += state.dir[1]
        return state
    state.dir = turns[(state.dir, action)]
    return state


def is_goal(state, goal):
    return state.x, state.y == goal[0], goal[1]


def UCS(start_state, goal_pos):
    pq = PriorityQueue((0, start_state))
    visited = set()
    while pq:
        curr_cost, curr_state = pq.get()
        visited.add((curr_state.x, curr_state.y))
        for action in actions:
            if valid(curr_state, action):
                next_state = update_state(curr_state, action)
                if is_goal(next_state, goal_pos):
                    return curr_cost + 1
                if (next_state.x, next_state.y) not in visited:
                    cost = curr_cost + action[1]
                    pq.put((cost, curr_state))


if __name__ == "__main__":
    grid = get_map("test.txt")
    start_pos, goal_pos = get_start_and_goal(grid)
    start_state = State(*start_pos, E)
    path_cost = UCS(start_state, goal_pos)

File: day16/aoc16.cpp
#include <fstream>
#include <iostream>
#include <map>
#include <set>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

#include "pq.h"
#include "utils.h"

// Globals
Map map;

Dir N = Dir(-1, 0);
Dir S = Dir(1, 0);
Dir E = Dir(0, 1);
Dir W = Dir(0, -1);

std::vector<Dir> dirs = {N, S, E, W};

Action L = Action('L', 1000);
Action R = Action('R', 1000);
Action F = Action('F', 1);

std::vector<Action> actions = {L, R, F};

std::map<std::pair<Dir, Action>, Dir> turns = {
    {{N, L}, W}, {{N, R}, E}, {{S, L}, E}, {{S, R}, W},
    {{E, L}, N}, {{E, R}, S}, {{W, L}, S}, {{W, R}, N},
};

void getMap(const std::string &path, Map &map, Pos &start_pos, Pos &end_pos) {
  std::ifstream input(path);
  std::string line;
  if (!input.is_open()) {
    throw std::runtime_error("Could not open file: " + path);
  }
  int x_val = 0;
  while (std::getline(input, line)) {
    std::vector<char> currLine;
    int y_val = 0;
    for (char ch : line) {
      currLine.push_back(ch);
      if (ch == 'S') {
        start_pos.x = x_val;
        start_pos.y = y_val;
      } else if (ch == 'E') {
        end_pos.x = x_val;
        end_pos.y = y_val;
      }
    }
    map.push_back(currLine);
    x_val++;
  }
}

bool clear(const State &state, const Dir &dir) {
  return map[state.pos.x + dir.x][state.pos.y + dir.y] == '.';
}

bool valid(const Action &action, const State &state) {
  if (action == F) {
    return clear(state, state.dir);
  }
  return clear(state, turns[{state.dir, action}]);
}

State updateState(State &state, const Action &action) {
  if (action == F) {
    int x = state.pos.x + state.dir.x;
    int y = state.pos.y + state.dir.y;
    return State(Pos(x, y), state.dir);
  }
  return State(state.pos, turns[{state.dir, action}]);
}

bool isGoal(State s, Pos end_pos) {
  int x = s.pos.x;
  int y = s.pos.y;
  return (x == end_pos.x) && (y == end_pos.y);
}

int UCS(Pos &start_pos, Pos end_pos) {
  State currState = State(start_pos, E);
  std::set<Pos> visited;
  PQ pq;
  pq.push(currState, 0);

  while (!pq.empty()) {
    auto [state, cost] = pq.pop();
    visited.insert(state.pos);
    for (Action action : actions) {
      if (valid(action, state)) {
        State next_state = updateState(state, action);
        if (isGoal(next_state, end_pos)) {
          return cost + 1;
        }
        if (!visited.contains(state.pos)) {
          int next_cost = cost += action.cost;
          pq.push(next_state, next_cost);
        }
      }
    }
  }
  return -1;
}

int main() {
  std::string path = "test.txt";
  Pos start_pos;
  Pos end_pos;

  getMap(path, map, start_pos, end_pos);
  int cost = UCS(start_pos, end_pos);
  std::cout << "Minimal path cost: " << cost << std::endl;

  return 0;
}
